Remove resource_set limitations

agomezl opened the pull request:
Currently, the resource_set attribute of `ctx.actions.run` (an similar) can only be set to a function returning a dictionary with a limited set of keys/resources (mem, cpu, and local_test). This change removes this restriction, allowing any resource known to the resource manager.

RELNOTES: `resource_set` function can now reference custom resources

Closes #27950

fmeum commented on src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java:
```diff
@@ -914,22 +912,29 @@ public ResourceSet buildResourceSet(OS os, int inputsSize)
-          throw new EvalException(message);
+        ImmutableMap.Builder<String, Double> resourceSetBuilder = ImmutableMap.builder();
+        for (String key : resourceSetMapRaw.keySet()) {
+          if (!validResources.contains(key)) {
```
`validResources` should probably be renamed now that other resources are also valid

agomezl replied:
It is now renamed to `defaultResources` 👍🏼

agomezl commented:
I don't know what is the deal with the `macos-arm64` tests, but they also fail in my personal mac with:

```
ERROR: Error computing the main repository mapping: Error accessing registry https://bcr.bazel.build/: 
Failed to fetch registry file https://bcr.bazel.build/modules/rules_cc/0.2.14/MODULE.bazel: No route to host
```

fmeum commented:
@gregestren Friendly ping, although I assume this has some overlap with local exec (@meisterT for your opinion on this)

agomezl commented on src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java:
```diff
@@ -914,22 +912,29 @@ public ResourceSet buildResourceSet(OS os, int inputsSize)
+            resourceSetBuilder.buildOrThrow(),
             (int)
                 getNumericOrDefault(
                     resourceSetMapRaw, "local_test", DEFAULT_RESOURCE_SET.getLocalTestCount()));
```
I suspect it is because `local_test` as a resource is "nothing special" an just whatever value [--local_test_jobs](https://bazel.build/reference/command-line-reference#build-flag--local_test_jobs) has. Actually, it seems like `--local_test_jobs=n == --local_resource=local_test=n` with some added precedence 🤷🏼. We could just get rid of it completely.

agomezl committed:
Remove resource_set limitations

Currently, the resource_set attribute of `ctx.actions.run` (an similar)
can only be set to a function returning a dictionary with a limited set
of keys/resources (mem, cpu, and local_test). This change removes this
restriction, allowing any resource know to the resource manager.

RELNOTES: `resource_set` function can now reference custom resources

Closes #27950

agomezl committed:
Fix how local_test is routed into the ResourceSet constructor

agomezl committed:
Add tests to StarlarkRuleContextTest.java

agomezl committed:
Tweak resource_set documentation

agomezl committed:
Update the documentation of --local_resources

agomezl committed:
Add documentation

agomezl committed:
Create a ResourceCollector class

agomezl committed:
Fix tests

agomezl committed:
Simplify things a bit more

agomezl commented:
@gregestren Could we give this change another go? 😆.

I have updated it to account for #30263 and #29285, tried to share some code between the dictionary and function variants.
